oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java

```diff
@@ -0,0 +1,103 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.query;
+
+import java.util.Locale;
+
+/**
+ * Formatter for JCR queries in order to make them easier to read. Formatting is
+ * done on a best-effort basis.
+ * 
+ * Warning: formatting is also done within e.g. string literals. So there is no
+ * guarantee that the formatted query is semantically equal to the original one!
+ */
+public class QueryFormatter {
+
+    /**
+     * Detect whether the query is an XPath query.
+     * 
+     * @param query the query
+     * @param language the language, if known, or null
+     * @return true if xpath
+     */
+    public static boolean isXPath(String query, String language) {
+        if (language != null) {
+            return "xpath".equals(language);
+        }
+        query = query.trim().toLowerCase(Locale.ENGLISH);
+        // explain queries
+        if (query.startsWith("explain")) {
+            query = query.substring("explain".length()).trim();
+            if (query.startsWith("measure")) {
+                query = query.substring("explain".length()).trim();
+            }            
+        }
+        // union queries
+        while (query.startsWith("(")) {
+            query = query.substring("(".length()).trim();
+        }
+        if (query.startsWith("select")) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Format the query into a more human-readable way, by adding newlines.
+     * Warning: newlines are also added inside e.g. string literals.
+     * 
+     * @param query the query (may not be null)
+     * @param language the query language, or null if unknown
+     * @return the formatted query
+     */
+    public static String format(String query, String language) {
+        boolean xpath = isXPath(query, language);
+        if (xpath) {
+            return formatXPath(query);
+        } else {
+            return formatSQL(query);
+        }
+    }
+
+    private static String formatXPath(String query) {
+        query = query.replaceAll("\\[", "\\[\n  ");
+        for (String term : new String[] {
+                "and ", "or ", "order by ", "option\\("
+        }) {
+            // xpath is case sensitive
+            query = query.replaceAll(" (" + term + ")", "\n  $1");
+        }
+        // remove duplicate newlines
+        query = query.replaceAll("\n+", "\n");
+        return query;
+    }
+    
+    private static String formatSQL(String query) {
+        for (String term : new String[] {
+                "union ", "from ", "where ", "and ", "or ", "order by ", "option\\("
+        }) {
+            // SQL is case insensitive, so we use (?i)
+            query = query.replaceAll("(?i) (" + term + ")", "\n  $1");
+        }
+        // remove duplicate newlines
+        query = query.replaceAll("\n+", "\n");
+        return query;
+    }
```

Comment:
Some of these keywords are common enough that they will likely appear in the text of some queries, which will then result in line breaks being added in the wrong places. Would it be very complicated to change the logic to detect if they words are used as sql keywords or not? Maybe something with checking if the word appears inside quotes or not? Or maybe it's ok for the output to sometimes be wrongly formatted, as this is mostly a visual aid.

Reply:
Yes this is a concern, if the output is then run as a query... (By the human or automatically).
I was thinking on how to solve it, but I don't think there is a simple solution that would cover all cases.
Maybe we could add a comment in the output, but that's ugly as well...

The output is only used in one place: LOG.debug, so I think the risk is not all that great now.
My recommendation is to just accept that risk.
But not use the code anywhere else, that's why I added the warning in the Javadoc.